<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.9.1"/>
<title>XDPP1100 Firmware: cmsis_os_exceptions.h File Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="stylesheet.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectlogo"><img alt="Logo" src="logo_infineon.jpg"/></td>
  <td style="padding-left: 0.5em;">
   <div id="projectname">XDPP1100 Firmware
   </div>
   <div id="projectbrief">The firmware documentation for the XDPP1100 device family.</div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.9.1 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('cmsis__os__exceptions_8h.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#enum-members">Enumerations</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">cmsis_os_exceptions.h File Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p>Library for using a simple Structured Exception Handling technique in an CMSIS-RTOS environment with multiple threads.  
<a href="#details">More...</a></p>
<div class="textblock"><code>#include &quot;<a class="el" href="cmsis__os_8h.html">cmsis_os.h</a>&quot;</code><br />
<code>#include &quot;<a class="el" href="linked__list_8h.html">linked_list.h</a>&quot;</code><br />
<code>#include &lt;setjmp.h&gt;</code><br />
</div><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structos_exception__t.html">osException_t</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:a9c2d6562201b8a9a10a242d6b4e75fad"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">THROW</a>(exc_type)&#160;&#160;&#160;<a class="el" href="cmsis__os__exceptions_8h.html#acfdf2ada84293c172704044f7876eb03">osExcThrow</a>((<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a>)exc_type, <a class="el" href="environment_8h.html#a3c4b9b00d2865ad1d07c0c551f9b67c1">PROGRAM_COUNTER</a>)</td></tr>
<tr class="separator:a9c2d6562201b8a9a10a242d6b4e75fad"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7bc5cf3ad6ddce8de80fbe0dae4b8386"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a7bc5cf3ad6ddce8de80fbe0dae4b8386">THROW_IF</a>(cond,  exc_type)&#160;&#160;&#160;if (cond) <a class="el" href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">THROW</a>(exc_type)</td></tr>
<tr class="separator:a7bc5cf3ad6ddce8de80fbe0dae4b8386"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a58c579d2d0639c481f6f404d20de00e9"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a58c579d2d0639c481f6f404d20de00e9">THROW_OTHER</a>(exc_type)</td></tr>
<tr class="separator:a58c579d2d0639c481f6f404d20de00e9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad2746371528bdf15c3910b7bf217dac0"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#ad2746371528bdf15c3910b7bf217dac0">TRY</a></td></tr>
<tr class="separator:ad2746371528bdf15c3910b7bf217dac0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aefda3c41e567de5717e579212109a355"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#aefda3c41e567de5717e579212109a355">TRY_AGAIN</a></td></tr>
<tr class="separator:aefda3c41e567de5717e579212109a355"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0e92453d34d1a100d58c6177fd7747fb"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a0e92453d34d1a100d58c6177fd7747fb">CATCH</a>(exc_type)</td></tr>
<tr class="separator:a0e92453d34d1a100d58c6177fd7747fb"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aad867ef569869bd836e12dcba471aa95"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#aad867ef569869bd836e12dcba471aa95">CATCH_OTHERS</a></td></tr>
<tr class="separator:aad867ef569869bd836e12dcba471aa95"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0e2a75478cd44f1666a6aca626c5c50b"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a0e2a75478cd44f1666a6aca626c5c50b">FINALLY</a></td></tr>
<tr class="separator:a0e2a75478cd44f1666a6aca626c5c50b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a42527369fadf3dcd985f6b58160284f1"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a42527369fadf3dcd985f6b58160284f1">TRY_END</a></td></tr>
<tr class="separator:a42527369fadf3dcd985f6b58160284f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8e5fdcdfe948aeed8558f0d81392e941"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a8e5fdcdfe948aeed8558f0d81392e941">EXCEPTION</a>&#160;&#160;&#160;exc_catcher-&gt;last_exception</td></tr>
<tr class="separator:a8e5fdcdfe948aeed8558f0d81392e941"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="enum-members"></a>
Enumerations</h2></td></tr>
<tr class="memitem:acd9f76a1fbd8bc9084ff34add637094f"><td class="memItemLeft" align="right" valign="top">enum &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a> { <br />
&#160;&#160;<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a> = 0, 
<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fa7dcca268506a6a937762fa24092b2894">ILLEGAL_PARAMETER</a>, 
<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094faad9dee005a3d0f9137b2ac1e0869f89b">TIMEOUT</a>, 
<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fa0132f16d17c27d46245f9361645fe5e4">OUT_OF_RESOURCE</a>, 
<br />
&#160;&#160;<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fa0d80fe80390ba8d708c0cb2fa707f08d">FAILED_COMMUNICATION</a>, 
<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fab1cbc05a88cbd1e46eae408900283edf">CORRUPT_MEMORY</a>, 
<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094faa961f0f22dd569121e4d9411ec9cd4e4">SYSTEM_FAULT</a>
<br />
 }</td></tr>
<tr class="separator:acd9f76a1fbd8bc9084ff34add637094f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:a5b7db92928c474b4e385d56ede1c0cec"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a5b7db92928c474b4e385d56ede1c0cec">osExcInitialize</a> (<a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a> *(*<a class="el" href="cmsis__os__exceptions_8c.html#aa7e32e09829248466aaba2fed171bb25">exc_catcher_get</a>)(const <a class="el" href="cmsis__os_8h.html#adfeb153a84a81309e2d958268197617f">osThreadId</a> thread_id))</td></tr>
<tr class="separator:a5b7db92928c474b4e385d56ede1c0cec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a76af7a7f62626847bc684d2b2bde9373"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#a76af7a7f62626847bc684d2b2bde9373">osExcGuardSet</a> (<a class="el" href="linked__list_8h.html#ae5149af60f091aa9736e470752558e5b">Linked_Node_t</a> *guard_node)</td></tr>
<tr class="separator:a76af7a7f62626847bc684d2b2bde9373"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad6f1875324dff93506f5d8338b92b781"><td class="memItemLeft" align="right" valign="top">jmp_buf *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#ad6f1875324dff93506f5d8338b92b781">osExcLastGuardPop</a> (<a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a> *exc_catcher)</td></tr>
<tr class="separator:ad6f1875324dff93506f5d8338b92b781"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:acfdf2ada84293c172704044f7876eb03"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="cmsis__os__exceptions_8h.html#acfdf2ada84293c172704044f7876eb03">osExcThrow</a> (<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a> type, uint32_t program_counter)</td></tr>
<tr class="separator:acfdf2ada84293c172704044f7876eb03"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Library for using a simple Structured Exception Handling technique in an CMSIS-RTOS environment with multiple threads. </p>
</div><h2 class="groupheader">Macro Definition Documentation</h2>
<a class="anchor" id="a0e92453d34d1a100d58c6177fd7747fb"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define CATCH</td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">exc_type</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line"><span class="keywordflow">break</span>; \</div>
<div class="line">        case exc_type: \</div>
<div class="line">            uncaught_exception = <a class="code" href="fw__bool_8h.html#ab2da649e4cae61053af4b4fbb6469e3d">boolean_false</a>;</div>
<div class="ttc" id="fw__bool_8h_html_ab2da649e4cae61053af4b4fbb6469e3d"><div class="ttname"><a href="fw__bool_8h.html#ab2da649e4cae61053af4b4fbb6469e3d">boolean_false</a></div><div class="ttdeci">#define boolean_false</div><div class="ttdef"><b>Definition:</b> fw_bool.h:35</div></div>
</div><!-- fragment --><p>Block in which a specific exception is caught and handled. </p>

</div>
</div>
<a class="anchor" id="aad867ef569869bd836e12dcba471aa95"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define CATCH_OTHERS</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line"><span class="keywordflow">break</span>; \</div>
<div class="line">        } <span class="comment">/* end of for(;;) */</span> \</div>
<div class="line">        break; <span class="comment">/* break statement to prevent fall-through into default from each case */</span> \</div>
<div class="line">        default: \</div>
<div class="line">            uncaught_exception = <a class="code" href="fw__bool_8h.html#ab2da649e4cae61053af4b4fbb6469e3d">boolean_false</a>;</div>
<div class="ttc" id="fw__bool_8h_html_ab2da649e4cae61053af4b4fbb6469e3d"><div class="ttname"><a href="fw__bool_8h.html#ab2da649e4cae61053af4b4fbb6469e3d">boolean_false</a></div><div class="ttdeci">#define boolean_false</div><div class="ttdef"><b>Definition:</b> fw_bool.h:35</div></div>
</div><!-- fragment --><p>Instead of using FINALLY all other exceptions can be caught with this macro. CATCH_OTHERS is intended to definitely handle the exception and not to THROW it further. </p>

</div>
</div>
<a class="anchor" id="a8e5fdcdfe948aeed8558f0d81392e941"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define EXCEPTION&#160;&#160;&#160;exc_catcher-&gt;last_exception</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The data structure of the last exception that was caught within the current TRY block. Contains the type and additional information (see Exception_t). </p>

</div>
</div>
<a class="anchor" id="a0e2a75478cd44f1666a6aca626c5c50b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define FINALLY</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line"><span class="keywordflow">break</span>; \</div>
<div class="line">        } <span class="comment">/* end of for(;;) */</span> \</div>
<div class="line">        <span class="comment">/* no break statement for fall-through into default from each case */</span> \</div>
<div class="line">        <span class="comment">/*lint -fallthrough */</span> \</div>
<div class="line">        default:</div>
</div><!-- fragment --><p>Instructions in this block are always executed before an uncaught exception is thrown further up. </p>

</div>
</div>
<a class="anchor" id="a9c2d6562201b8a9a10a242d6b4e75fad"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define THROW</td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">exc_type</td><td>)</td>
          <td>&#160;&#160;&#160;<a class="el" href="cmsis__os__exceptions_8h.html#acfdf2ada84293c172704044f7876eb03">osExcThrow</a>((<a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a>)exc_type, <a class="el" href="environment_8h.html#a3c4b9b00d2865ad1d07c0c551f9b67c1">PROGRAM_COUNTER</a>)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Throws an exception. The macro may be used everywhere since no action is taken if there is no TRY-CATCH-block to guard the THROW or even if the thread was not registered.<br />
 Exceptions can only be handled in code that is guarded by a TRY-CATCH within a registered thread. Note: PROGRAM_COUNTER is compiler specific and has to be defined elsewhere. </p>

</div>
</div>
<a class="anchor" id="a7bc5cf3ad6ddce8de80fbe0dae4b8386"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define THROW_IF</td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">cond, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">exc_type&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td>&#160;&#160;&#160;if (cond) <a class="el" href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">THROW</a>(exc_type)</td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Readability macro for a single-line conditional THROW which puts the intention to throw an exception at the beginning of the line. </p>

</div>
</div>
<a class="anchor" id="a58c579d2d0639c481f6f404d20de00e9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define THROW_OTHER</td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname">exc_type</td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line">uncaught_exception = <a class="code" href="fw__bool_8h.html#ac28db6ba61eab23fff2e0d7404bc9805">boolean_true</a>; <span class="comment">/* Although exception was caught throw another one after FINALLY */</span> \</div>
<div class="line">    exc_catcher-&gt;last_exception.type = exc_type;</div>
<div class="ttc" id="fw__bool_8h_html_ac28db6ba61eab23fff2e0d7404bc9805"><div class="ttname"><a href="fw__bool_8h.html#ac28db6ba61eab23fff2e0d7404bc9805">boolean_true</a></div><div class="ttdeci">#define boolean_true</div><div class="ttdef"><b>Definition:</b> fw_bool.h:34</div></div>
</div><!-- fragment --><p>Used within CATCH for changing the exception and throwing it further up.<br />
 Note that the unintended use of THROW instead of THROW_OTHER within CATCH would bypass the FINALLY block. </p>

</div>
</div>
<a class="anchor" id="ad2746371528bdf15c3910b7bf217dac0"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define TRY</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line">{ \</div>
<div class="line">    jmp_buf guard; <a class="code" href="linked__list_8h.html#ae5149af60f091aa9736e470752558e5b">\</a></div>
<div class="line"><a class="code" href="linked__list_8h.html#ae5149af60f091aa9736e470752558e5b">    Linked_Node_t</a> guard_node = {(<span class="keywordtype">void</span>*) guard, NULL, NULL}; \</div>
<div class="line">    osExceptionCatcher_t* exc_catcher = <a class="code" href="cmsis__os__exceptions_8h.html#a76af7a7f62626847bc684d2b2bde9373">osExcGuardSet</a>(&amp;guard_node); \</div>
<div class="line">    exc_catcher-&gt;last_exception.type = (<a class="code" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a>) setjmp(guard); <a class="code" href="fw__bool_8h.html#a9201c1178e191a87a38dd59c7fc36ff6">\</a></div>
<div class="line"><a class="code" href="fw__bool_8h.html#a9201c1178e191a87a38dd59c7fc36ff6">    boolean</a> uncaught_exception = (exc_catcher-&gt;last_exception.type != <a class="code" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a>); \</div>
<div class="line">    switch(exc_catcher-&gt;last_exception.type) \</div>
<div class="line">    { \</div>
<div class="line">        case <a class="code" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a>: <span class="comment">/* normal operation */</span> \</div>
<div class="line">        for(;;) <span class="comment">/* following breaks exit this loop -&gt; default is used after each case */</span> \</div>
<div class="line">        {</div>
<div class="ttc" id="linked__list_8h_html_ae5149af60f091aa9736e470752558e5b"><div class="ttname"><a href="linked__list_8h.html#ae5149af60f091aa9736e470752558e5b">Linked_Node_t</a></div><div class="ttdeci">struct Linked_Node Linked_Node_t</div><div class="ttdef"><b>Definition:</b> linked_list.h:30</div></div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_acd9f76a1fbd8bc9084ff34add637094f"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a></div><div class="ttdeci">osError_t</div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.h:344</div></div>
<div class="ttc" id="fw__bool_8h_html_a9201c1178e191a87a38dd59c7fc36ff6"><div class="ttname"><a href="fw__bool_8h.html#a9201c1178e191a87a38dd59c7fc36ff6">boolean</a></div><div class="ttdeci">#define boolean</div><div class="ttdef"><b>Definition:</b> fw_bool.h:32</div></div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_a76af7a7f62626847bc684d2b2bde9373"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#a76af7a7f62626847bc684d2b2bde9373">osExcGuardSet</a></div><div class="ttdeci">osExceptionCatcher_t * osExcGuardSet(Linked_Node_t *guard_node)</div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.c:70</div></div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a></div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.h:348</div></div>
</div><!-- fragment --><p>Sets up a try-guard in which one can execute a function that might throw an exception. A minimum top-level TRY block looks like this: </p><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;TRY</div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;    something_risky();</div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;CATCH_OTHERS</div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;    handle_exceptions();</div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;TRY_END</div>
</div><!-- fragment --> 
</div>
</div>
<a class="anchor" id="aefda3c41e567de5717e579212109a355"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define TRY_AGAIN</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line"><span class="comment">/* Loop until an exception occurs. */</span> <a class="code" href="cmsis__os__exceptions_8h.html#ad2746371528bdf15c3910b7bf217dac0">\</a></div>
<div class="line"><a class="code" href="cmsis__os__exceptions_8h.html#ad2746371528bdf15c3910b7bf217dac0">    TRY</a> <span class="keywordflow">for</span> (uint8_t always_true = 1; always_true; )</div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_ad2746371528bdf15c3910b7bf217dac0"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#ad2746371528bdf15c3910b7bf217dac0">TRY</a></div><div class="ttdeci">#define TRY</div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.h:454</div></div>
</div><!-- fragment --><p>This macro is needed for the use case of an efficient TRY block within a thread's forever loop. The utilization looks like this: </p><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;for (;;)</div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;{</div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;    TRY_AGAIN</div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;        something_risky();</div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;    CATCH_OTHERS</div>
<div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;        handle_exceptions();</div>
<div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;    TRY_END</div>
<div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;}</div>
</div><!-- fragment --><p> <br />
 Usually one would think to use the following code: </p><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;for (;;)</div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;{</div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;    TRY</div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;        something_risky();</div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;    CATCH_OTHERS</div>
<div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;        handle_exceptions();</div>
<div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;    TRY_END</div>
<div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;}</div>
</div><!-- fragment --><p> This has the disadvantage that the time costly setup of a TRY block would be done in each iteration. So an inner loop is needed which will be exited in case of an exception. The thread will then continue with the outer loop. With this construction the TRY block does not have to be set up for each iteration.<br />
 So then why not use these lines?: </p><div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;for (;;)</div>
<div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;{</div>
<div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;    TRY</div>
<div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;        for (;;)</div>
<div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;        {</div>
<div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;            something_risky();</div>
<div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;        }</div>
<div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;    CATCH_OTHERS</div>
<div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;        handle_exceptions();</div>
<div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;    TRY_END</div>
<div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;}</div>
</div><!-- fragment --><p> In this case the compiler will throw a warning for unreachable code of the <a class="el" href="cmsis__os__exceptions_8h.html#a0e92453d34d1a100d58c6177fd7747fb">CATCH(_OTHERS)</a>, FINALLY and TRY_END macros. This warning is unnecessary since longjmp within THROW will exit the inner forever loop in the TRY block.<br />
 To inhibit the warning a variable in the forever loop of TRY_AGAIN is used. This will also prevent the compiler from removing the "unreachable code sections" in a possible optimization step. </p>

</div>
</div>
<a class="anchor" id="a42527369fadf3dcd985f6b58160284f1"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define TRY_END</td>
        </tr>
      </table>
</div><div class="memdoc">
<b>Value:</b><div class="fragment"><div class="line"><span class="keywordflow">break</span>; \</div>
<div class="line">    } <span class="comment">/* end of switch */</span> \</div>
<div class="line">    <span class="comment">/*lint -e(774) Boolean within &#39;if&#39; always evaluates to False . */</span> \</div>
<div class="line">    if (exc_catcher-&gt;last_exception.type == <a class="code" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a>) \</div>
<div class="line">    { \</div>
<div class="line">        <span class="comment">/*lint -e(534) Suppress &quot;Ignoring return value&quot;. The returned guard is only used within THROW. */</span> <a class="code" href="cmsis__os__exceptions_8h.html#ad6f1875324dff93506f5d8338b92b781">\</a></div>
<div class="line"><a class="code" href="cmsis__os__exceptions_8h.html#ad6f1875324dff93506f5d8338b92b781">        osExcLastGuardPop</a>(exc_catcher); \</div>
<div class="line">    } \</div>
<div class="line">    <span class="comment">/*lint -e(774) Boolean within &#39;if&#39; always evaluates to False . */</span> \</div>
<div class="line">    else <span class="keywordflow">if</span> (uncaught_exception) \</div>
<div class="line">    { <a class="code" href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">\</a></div>
<div class="line"><a class="code" href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">        THROW</a>(exc_catcher-&gt;last_exception.type); \</div>
<div class="line">    } \</div>
<div class="line">    else <span class="comment">/* (There was an exception &amp;&amp; it was caught) */</span> \</div>
<div class="line">    { \</div>
<div class="line">        <span class="comment">/* At this point last_exception.type is always unequal to NO_ERROR. For the TRY block</span></div>
<div class="line"><span class="comment">         * above it must look like as if there was no exception and that its TRY_END must remove</span></div>
<div class="line"><span class="comment">         * its guard. */</span> \</div>
<div class="line">        exc_catcher-&gt;last_exception.type = <a class="code" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a>; \</div>
<div class="line">    } \</div>
<div class="line">}</div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_ad6f1875324dff93506f5d8338b92b781"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#ad6f1875324dff93506f5d8338b92b781">osExcLastGuardPop</a></div><div class="ttdeci">jmp_buf * osExcLastGuardPop(osExceptionCatcher_t *exc_catcher)</div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.c:84</div></div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_a9c2d6562201b8a9a10a242d6b4e75fad"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#a9c2d6562201b8a9a10a242d6b4e75fad">THROW</a></div><div class="ttdeci">#define THROW(exc_type)</div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.h:426</div></div>
<div class="ttc" id="cmsis__os__exceptions_8h_html_acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94"><div class="ttname"><a href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94">NO_ERROR</a></div><div class="ttdef"><b>Definition:</b> cmsis_os_exceptions.h:348</div></div>
</div><!-- fragment --><p>Closes the guarded try-block by removing the guard or throwing an uncaught exception further up. </p>

</div>
</div>
<h2 class="groupheader">Enumeration Type Documentation</h2>
<a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">enum <a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>The error classes are indicating the original cause of an exception. </p><dl class="section note"><dt>Note</dt><dd>A detected error during runtime is defined as an exception. The exception is usually then used to interrupt the normal program flow. </dd>
<dd>
The classes are sorted by their estimated likelihood to occur. </dd></dl>
<table class="fieldtable">
<tr><th colspan="2">Enumerator</th></tr><tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094fabf350750d0d4fabd8954c0f1e9bbae94"></a>NO_ERROR&#160;</td><td class="fielddoc">
<p>The return value of setjmp in case no exception was thrown due to an error (= setjmp was not reached through longjmp). </p>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094fa7dcca268506a6a937762fa24092b2894"></a>ILLEGAL_PARAMETER&#160;</td><td class="fielddoc">
<ul>
<li>
A function parameter is outside its permitted range. Using it could lead to a null pointer access or the access of a wrong memory area like an array index that is out of bounds or a non-existent function. </li>
</ul>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094faad9dee005a3d0f9137b2ac1e0869f89b"></a>TIMEOUT&#160;</td><td class="fielddoc">
<ul>
<li>
An event (interrupt, change of an interface value) did not occur during a given time frame or number of polling attempts. </li>
</ul>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094fa0132f16d17c27d46245f9361645fe5e4"></a>OUT_OF_RESOURCE&#160;</td><td class="fielddoc">
<ul>
<li>
During the attempt to allocate a HW resource like a block of memory or a special function peripheral the dedicated FW module detects that there is no resource remaining. </li>
<li>
A stack overflow was detected for the currently running task (CMSIS: OS_ERR_STK_OVF). </li>
<li>
The pending ISR calls (new IRQ, semaphore, queues) are exceeded (CMSIS: OS_ERR_FIFO_OVF). </li>
<li>
A mailbox (not message box) overflow was detected (CMSIS: OS_ERR_MBX_OVF). </li>
</ul>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094fa0d80fe80390ba8d708c0cb2fa707f08d"></a>FAILED_COMMUNICATION&#160;</td><td class="fielddoc">
<ul>
<li>
The protocol of a communication was violated (e.g. wrong direction, wrong number of data bytes). </li>
<li>
The redundancy check of the transferred data has failed. </li>
</ul>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094fab1cbc05a88cbd1e46eae408900283edf"></a>CORRUPT_MEMORY&#160;</td><td class="fielddoc">
<ul>
<li>
A redundancy check has failed for the data stored in any kind of memory. </li>
</ul>
</td></tr>
<tr><td class="fieldname"><a class="anchor" id="acd9f76a1fbd8bc9084ff34add637094faa961f0f22dd569121e4d9411ec9cd4e4"></a>SYSTEM_FAULT&#160;</td><td class="fielddoc">
<ul>
<li>
All errors that cannot be detected by software fall into this class. </li>
</ul>
</td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a class="anchor" id="a76af7a7f62626847bc684d2b2bde9373"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a>* osExcGuardSet </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="linked__list_8h.html#ae5149af60f091aa9736e470752558e5b">Linked_Node_t</a> *&#160;</td>
          <td class="paramname"><em>guard_node</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Implicitly used within TRY macro (so not intended for explicit use!): Links a new guard (the back-jump address in case of an exception) to the exception catcher for the current thread. No dynamic memory allocation will be used here: A node containing the guard is expected by this function since the data for exception handling has to be created on the stack of the thread. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">guard_node</td><td>Linked list node containing the guard (will be modified) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Data object for managing the exception handling. </dd></dl>

</div>
</div>
<a class="anchor" id="a5b7db92928c474b4e385d56ede1c0cec"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void osExcInitialize </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a> *(*)(const <a class="el" href="cmsis__os_8h.html#adfeb153a84a81309e2d958268197617f">osThreadId</a> thread_id)&#160;</td>
          <td class="paramname"><em>exc_catcher_get</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Initializes the Structured Exception Handling to use it within an RTOS environment. Call this function with a NULL parameter in case this library needs to be utilized before <a class="el" href="cmsis__os_8h.html#aab668ffd2ea76bb0a77ab0ab385eaef2">osKernelStart()</a> and before the memory has been initialized. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">exc_catcher_get</td><td>Callback function which defines how the exception catchers of the threads can be accessed from the thread context </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a class="anchor" id="ad6f1875324dff93506f5d8338b92b781"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">jmp_buf* osExcLastGuardPop </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structos_exception_catcher__t.html">osExceptionCatcher_t</a> *&#160;</td>
          <td class="paramname"><em>exc_catcher</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Implicitly used within THROW and TRY_END macro (so not intended for explicit use!): Unlinks the last guard of the current thread from the Exception FW Module and returns it. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">exc_catcher</td><td>Exception catcher of the current thread </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The last guard of the indicated thread or NULL if the thread was not registered for exception handling. </dd></dl>

</div>
</div>
<a class="anchor" id="acfdf2ada84293c172704044f7876eb03"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void osExcThrow </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="cmsis__os__exceptions_8h.html#acd9f76a1fbd8bc9084ff34add637094f">osError_t</a>&#160;</td>
          <td class="paramname"><em>type</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">uint32_t&#160;</td>
          <td class="paramname"><em>program_counter</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Implicitly used within THROW macro (so not intended for explicit use!): Attaches additional information to the specified exception and throws it to the last guard of the current thread. </p><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">type</td><td>ID indicating the type of the exception </td></tr>
    <tr><td class="paramname">program_counter</td><td>Value of the PC register when the exception was thrown </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="dir_bdd9a5d540de89e9fe90efdfc6973a4f.html">common</a></li><li class="navelem"><a class="el" href="dir_d52f7f52f412a942de55fb463ecb0248.html">modules</a></li><li class="navelem"><a class="el" href="dir_0a7117d21e2c10a7c842bb9c28e7f065.html">extensions</a></li><li class="navelem"><a class="el" href="cmsis__os__exceptions_8h.html">cmsis_os_exceptions.h</a></li>
    <li class="footer">Generated on Wed Nov 8 2023 10:16:59 for XDPP1100 Firmware by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.9.1 </li>
  </ul>
</div>
</body>
</html>
